simulator/main.rs

```diff
@@ -67,6 +116,343 @@ enum Value {
     Blob(Vec<u8>),
 }
 
+impl ArbitraryFrom<Vec<&Value>> for Value {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Vec<&Value>) -> Self {
+        if t.is_empty() {
+            return Value::Null;
+        }
+
+        let index = rng.gen_range(0..t.len());
+        t[index].clone()
+    }
+}
+
+impl ArbitraryFrom<ColumnType> for Value {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &ColumnType) -> Self {
+        match t {
+            ColumnType::Integer => Value::Integer(rng.gen_range(i64::MIN..i64::MAX)),
+            ColumnType::Float => Value::Float(rng.gen_range(-1e10..1e10)),
+            ColumnType::Text => Value::Text(gen_random_text(rng)),
+            ColumnType::Blob => Value::Blob(gen_random_text(rng).as_bytes().to_vec()),
+        }
+    }
+}
+
+struct LTValue(Value);
+
+impl ArbitraryFrom<Vec<&Value>> for LTValue {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Vec<&Value>) -> Self {
+        if t.is_empty() {
+            return LTValue(Value::Null);
+        }
+
+        let index = rng.gen_range(0..t.len());
+        LTValue::arbitrary_from(rng, t[index])
+    }
+}
+
+impl ArbitraryFrom<Value> for LTValue {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Value) -> Self {
+        match t {
+            Value::Integer(i) => LTValue(Value::Integer(rng.gen_range(i64::MIN..*i - 1))),
+            Value::Float(f) => LTValue(Value::Float(rng.gen_range(-1e10..*f - 1.0))),
+            Value::Text(t) => {
+                // Either shorten the string, or make at least one character smaller and mutate the rest
+                let mut t = t.clone();
+                if rng.gen_bool(0.01) {
+                    t.pop();
+                    LTValue(Value::Text(t))
+                } else {
+                    let index = rng.gen_range(0..t.len());
+                    let mut t = t.into_bytes();
+                    t[index] -= 1;
+                    // Mutate the rest of the string
+                    for i in (index + 1)..t.len() {
+                        t[i] = rng.gen_range(0..=255);
+                    }
+                    LTValue(Value::Text(String::from_utf8(t).unwrap()))
+                }
+            }
+            Value::Blob(b) => todo!(),
+            _ => unreachable!(),
+        }
+    }
+}
+
+struct GTValue(Value);
```

Comment:
What does gt mean

Reply:
The GT and LT in this context mean `GreaterThan` and `LessThan`. They're not real types, it's just newtype pattern because of the trait oriented generation mechanism. We can also switch to using explicit functions which could reduce the mental overhead.

A `GTValue` represents a `Value` `v2` that satisfies the `>` predicate with respect to another value `v1`. `GTValue` and `LTValue` don't have `Arbitrary` instances, but only `ArbitraryFrom<Value>`, in which they generate values that are either greater than, or less than the provided value.

Does that make sense?

Reply:
To be honest what I'm finding hard to wrap around in this PR is the move from simple functions to a whole bunch of abstract indirection that I'm still trying to parse. I don't say it's bad not at all. I'm just saying that it will take me some time to ensure this is what we want. Maybe you could provide me with some literature with regards to this kind of testing? I want to be sure I understand everything before merging because it seems interesting but I just don't want to merge something based on coolness. I also feel like we need more comments because there are a lot of abbreviations that make it hard to follow.

Reply:
agreed w/ more comments

Reply:
Yeah, I fully understand and agree with that. Let me try to address your points a bit.

- The indirections themselves might actually be useless for now, I've been using them because that's the style of writing I'm used to, so if they're creating extra mental load, we can drop them.
- The testing style itself comes from [QuickCheck](https://hackage.haskell.org/package/QuickCheck), the Property-Based Testing library from Haskell. The trait based generators are beneficial for implementing (1) higher order generator functions such as a generator that picks one of a few given options. You might've realized we use `rng.gen_range()` to do weighted random selection, QuickCheck has a `frequency` function that does that in a safer way, leveraing the typeclasses. The second benefit of the traits are that you can automatically derive properties from functions, given that the inputs themselves have generators defined for them(in other words, they implement `Arbitrary` trait). Here's a [Rust implementation of QuickCheck](https://github.com/BurntSushi/quickcheck)
- I also fully agree that we shouldn't add anything just because it looks fancy, I wouldn't have written the PR if I thought the main benefit was that. Let me state the contributions I see here:

1. I personally think the structured approach(creating a query type, randomly picking queries) scales better than the ad hoc approach in the current version. I have plans of adding more complex query types within the next week such as `Select .. Limit` or `OrderBy`.
2. It adds the `where` clauses to Select and Delete queries, and randomly generates those clauses. 

In my opinion, much of the complexity of the code comes from the second part. Generation of such queries is (I think) inherently complex, because generating an expression that satisfies a constraint is complex. I agree that they could be better documented, both by inline comments and a separate documentation discussing the generation strategies and properties.

What do you think? What should the forward steps be?

Reply:
Yeah a lot of this IMO hinges on whether we want the creation of arbitrary queries to be part of the simulator. If yes, then I see this kind of property-testing setup as a big plus cuz instead of having a bunch of if-else soup you can have a `Query::arbitrary()` type thing and encapsulate whatever "arbitrary" means inside those trait implementations. One big reason this PR has more complexity than `main` is that it's just doing more complex things than `main` is, as you said.

Reply:
I'm getting around these properties let's see how it goes but let's first try to clean up some things. If you can leave comments and improve naming of things that would help.